iwyu.cc

```diff
@@ -2560,7 +2553,24 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   }
 
   void ReportTypeUseInternal(SourceLocation used_loc, const Type* type,
-                             const set<const Type*>& blocked_types) {
+                             const set<const Type*>& blocked_types,
+                             UseKind use_kind) {
+    type = Desugar(type);
+    if (use_kind == UseKind::RemoveRefs ||
+        use_kind == UseKind::RemoveRefsAndPtr) {
+      if (const auto* ref_type = dyn_cast_or_null<ReferenceType>(type))
+        type = ref_type->getPointeeTypeAsWritten().getTypePtr();
+      // Keep use_kind to remove possible additional references in type alias
+      // chain.
+      // TODO(bolshakov): shouldn't Desugar be called here and below as well?
+    }
+    if (use_kind == UseKind::RemoveRefsAndPtr) {
+      if (const auto* ptr_type = dyn_cast_or_null<PointerType>(type)) {
```

Comment:
... but here you should be able to get away with `type->isPointerType()`. But maybe you want to keep them both as casts for symmetry?

Ah, also `dyn_cast_or_null`. Yeah, nevermind.